scipy/optimize/_zeros_py.py

```diff
@@ -1393,3 +1394,222 @@ def toms748(f, a, b, args=(), k=1,
+    As written in [Sachs2015]_ which in turn is based on Chandrupatla's
+    algorithm as described in [Scherer2018]_.
+    This allows vector arguments for x0, x1, and args
+
+    Parameters
+    ----------
+
+    f : callable
+        The function whose zero is wanted. It must be a function of a
+        single variable of the form ``f(x,a,b,c...)``, where ``a,b,c...``
+        are extra arguments that can be passed in the `args` parameter.
+    x0 : float, sequence, or ndarray
+        An initial estimate of the zero that should be somewhere near the
+        actual zero. If not scalar, then `func` must be vectorized and return
+        a sequence or array of the same shape as its first argument.
+    x1 : float, optional
+        Another estimate of the zero that should be somewhere near the
+        actual zero.
+    xtol : float, optional
+        Tolerance (absolute) for termination.
+    rtol : float, optional
+        Tolerance (relative) for termination.
+    maxiter : int, optional
+        If convergence is not achieved in `maxiter` iterations, an error is
+        raised. Must be >= 0.
+    callback : callable, optional
+        An optional user-supplied function, called after each iteration.
+        Called as ``callback(xk)``, where ``xk`` is the current value of `x0`.
+
+    Returns
+    -------
+    result : OptimizeResult
+        The optimization result represented as an ``OptimizeResult`` object.
+        Important attributes are: ``x`` the solution array, ``success`` an 
+        array of Boolean flags indicating if the optimizer exited successfully,
+        ``fun``, the current function value, ``nfev``, the number of function 
+        evaluations, and ``nit``, the number of iterations. See 
+        `OptimizeResult` for a description of other attributes.
+
+    Notes
+    -----
+
+    References
+    ----------
+
+    .. [Sachs2015]
+       Sachs, Jason
+       "Ten Little Algorithms, Part 5: Quadratic Extremum Interpolation and
+       Chandrupatla's Method",
+       https://www.embeddedrelated.com/showarticle/855.php
+       
+    .. [Scherer2018]
+       Scherer, Philipp
+       *Computational physics: Simulation of classical and Quantum Systems*.
+       SPRINGER INTERNATIONAL PU, 2018. Ch. 6.1.7.3.
+       
+    See Also
+    --------
+    brentq, brenth, ridder, bisect, newton
+    root_scalar : interface to root solvers for scalar functions
+    
+    Examples
+    --------
+    >>> from scipy import optimize
+    >>> def f(x,a):
+    ...     return a-x*x
+
+    >>> k = np.arange(1,8)
+    >>> y = optimize.chandrupatla(f,0,3,args=(k,))
+    >>> y
+    [1. 1.41421356  1.73205081  2. 2.23606798  2.44948974  2.64575131]
+
+    >>> y = optimize.chandrupatla(f,0,3,args=(7.0,))
+    >>> y
+    2.6457513110645907
+    """
+
+    # Initialization
+    b = x0
+    a = x1
+    fa = f(a, *args)
+    fb = f(b, *args)
+
+    # Make sure we know the size of the result
+    shape = np.shape(fa)
+    assert shape == np.shape(fb)
```

Comment:
Here's something I'm a bit confused about, but I think it'll help finish it off if I can get the dimensions correct. If I do what you have above: 
```python
a, b, fa, fb = np.broadcast_arrays(a, b, fa, fb)
shape = a.shape
a, b, fa, fb = np.atleast_1d(a, b, fa, fb)
```
In the scalar case, this results in `shape == ()` and I can't figure out how to combine the scalar + array case in the loop. However, if I do: 
```python
a, b, fa, fb = np.broadcast_arrays(a, b, fa, fb)
a, b, fa, fb = np.atleast_1d(a, b, fa, fb)
shape = a.shape
```
Then I have `shape == (1,)` in the scalar case which makes it (I think) straightforward to use the array case to do something like `a2 = a[iqi]` for example where `iqi = True`. Let me know if that makes sense or not, I'm still struggling a little bit with figuring out the shapes we want things at various points in the algo.

Reply:
You might need both. I think the first version gives you what the shape of the output should be, so you'll want to store that and reshape the results to that at the end. The second version is more convenient to work with throughout the algorithm. But for that shape, you can always use `a.shape` because it won't change, right?
Does that make sense? Does it answer the question?